File: spigot/src/main/java/us/minevict/mvutil/spigot/MinevictusUtilsSpigot.java
package us.minevict.mvutil.spigot;

import co.aikar.commands.MessageType;
import co.aikar.commands.PaperCommandManager;
import io.lettuce.core.RedisClient;
import io.lettuce.core.RedisURI;
import java.io.File;
import java.io.IOException;
import java.util.Locale;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;
import java.util.logging.Logger;
import me.tom.sparse.spigot.chat.menu.ChatMenuAPI;
import org.bukkit.configuration.InvalidConfigurationException;
import org.bukkit.event.HandlerList;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;
import us.minevict.mvutil.common.MinevictusUtilsPlatform;
import us.minevict.mvutil.common.utils.SetupPlatformless;
import us.minevict.mvutil.spigot.hidden.CloseDatabaseListener;
import us.minevict.mvutil.spigot.hidden.MegaChunkSizes;
import us.minevict.mvutil.spigot.permissions.PermissionsManager;
import us.minevict.mvutil.spigot.utils.Functions;

/**
 * @since 0.1.0
 */
public class MinevictusUtilsSpigot
    extends JavaPlugin
    implements MinevictusUtilsPlatform {
  @NotNull
  private final PermissionsManager permissionsManager;
  private RedisClient redisClient;

  public MinevictusUtilsSpigot() {
    super();

    this.permissionsManager = new PermissionsManager(this);
  }

  @Override
  public void onLoad() {
    SetupPlatformless.setup(this);
  }

  @Override
  public void onEnable() {
    saveDefaultConfig();
    getConfig().options().copyDefaults(true);

    redisClient = RedisClient.create(RedisURI.builder()
        .withHost(getConfig().getString("redis-hostname"))
        .withPort(getConfig().getInt("redis-port"))
        .build());

    getServer().getPluginManager().registerEvents(permissionsManager, this);
    getServer().getPluginManager().registerEvents(CloseDatabaseListener.getInstance(), this);

    MegaChunkSizes.MEGA_CHUNK_SIZE = getConfig()
        .getInt("megachunk-size", MegaChunkSizes.MEGA_CHUNK_SIZE);
    MegaChunkSizes.MEGA_CHUNK_OFFSET_X = getConfig()
        .getInt("megachunk-offset-x", MegaChunkSizes.MEGA_CHUNK_OFFSET_X);
    MegaChunkSizes.MEGA_CHUNK_OFFSET_Z = getConfig()
        .getInt("megachunk-offset-z", MegaChunkSizes.MEGA_CHUNK_OFFSET_Z);

    if (getConfig().getBoolean("megachunk-random-offset-x")) {
      MegaChunkSizes.MEGA_CHUNK_OFFSET_X += ThreadLocalRandom.current().nextInt();
    }
    if (getConfig().getBoolean("megachunk-random-offset-z")) {
      MegaChunkSizes.MEGA_CHUNK_OFFSET_Z += ThreadLocalRandom.current().nextInt();
    }

    if (getServer().getPluginManager().isPluginEnabled("ProtocolLib")) {
      ChatMenuAPI.init(this);
    }
  }

  @Override
  public void onDisable() {
    ChatMenuAPI.disable();

    HandlerList.unregisterAll(this);
    redisClient.shutdown(2, 5, TimeUnit.SECONDS);
  }

  @Override
  @NotNull
  public Logger getPlatformLogger() {
    return getLogger();
  }

  /**
   * Gets the current instance of this plugin.
   *
   * @return current instance of this plugin.
   */
  @NotNull
  public static MinevictusUtilsSpigot getInstance() {
    return getPlugin(MinevictusUtilsSpigot.class);
  }

  /**
   * Gets the permissions manager for managing a dependent's permissions.
   *
   * @return {@link PermissionsManager} for this instance.
   */
  @NotNull
  public PermissionsManager getPermissionsManager() {
    return permissionsManager;
  }

  /**
   * Prepare a command manager.
   *
   * @param commandManager The command manager to prepare.
   * @return The command managed for use in chaining or the likes.
   * @since 3.3.0
   */
  @NotNull
  public PaperCommandManager prepareAcf(@NotNull PaperCommandManager commandManager) {
    // Why the FUCK is MessageType not an enum??
    for (var type : new MessageType[]{
        MessageType.ERROR,
        MessageType.HELP,
        MessageType.INFO,
        MessageType.SYNTAX
    }) {
      for (var colour : org.bukkit.ChatColor.values()) {
        commandManager.setFormat(type, colour.ordinal(), colour);
      }
    }

    var acfLangFile = new File(commandManager.getPlugin().getDataFolder(), "acf-lang-en.yml");
    try {
      Functions.copyResource(
          this,
          "acf-lang-en.yml",
          acfLangFile
      );

      commandManager.getLocales().loadYamlLanguageFile(acfLangFile, Locale.ENGLISH);
    } catch (IOException | InvalidConfigurationException ex) {
      getLogger().warning("Could not save & load ACF language file for "
          + commandManager.getPlugin().getDescription().getName());
      ex.printStackTrace();
    }

    return commandManager;
  }

  @Override
  @NotNull
  public RedisClient getRedis() {
    return redisClient;
  }
}

File: spigot/src/main/kotlin/us/minevict/mvutil/spigot/MinevictusUtilsSpigot.kt
/**
 * MV-Util
 * Copyright (C) 2020 Mariell Hoversholm, Nahuel Dolores
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package us.minevict.mvutil.spigot

import co.aikar.commands.MessageType
import co.aikar.commands.PaperCommandManager
import co.aikar.idb.Database
import org.apache.commons.pool2.impl.GenericObjectPoolConfig
import org.bukkit.event.HandlerList
import org.bukkit.plugin.Plugin
import org.bukkit.plugin.java.JavaPlugin
import redis.clients.jedis.Jedis
import redis.clients.jedis.JedisPool
import redis.clients.jedis.Protocol
import us.minevict.mvutil.common.MinevictusUtilsPlatform
import us.minevict.mvutil.spigot.ext.BukkitChatColor
import us.minevict.mvutil.spigot.ext.copyResource
import us.minevict.mvutil.spigot.ext.createHikariDatabase
import us.minevict.mvutil.spigot.funkychunk.MegaChunk
import us.minevict.mvutil.spigot.internal.SetupPlatformless
import java.io.File
import java.util.*
import java.util.concurrent.ThreadLocalRandom
import java.util.logging.Logger

/**
 * @since 5.0.0
 */
class MinevictusUtilsSpigot : JavaPlugin(), MinevictusUtilsPlatform<Plugin, PaperCommandManager> {
    override lateinit var redis: JedisPool
        private set
    override val platformLogger: Logger
        get() = logger

    override fun onLoad() {
        SetupPlatformless.setup(this)
    }

    override fun onEnable() {
        saveDefaultConfig()
        config.options().copyDefaults(true)

        redis = JedisPool(
            GenericObjectPoolConfig<Jedis>(),
            config.getString("redis-hostname"),
            config.getInt("redis-port"),
            Protocol.DEFAULT_TIMEOUT,
            config.getString("redis-password")?.takeUnless(String::isBlank)
        )

        MegaChunk.MEGA_CHUNK_SIZE = config.getInt("megachunk-size")
        MegaChunk.MEGA_CHUNK_OFFSET_X = config.getInt("megachunk-offset-x")
        MegaChunk.MEGA_CHUNK_OFFSET_Z = config.getInt("megachunk-offset-z")

        if (config.getBoolean("megachunk-random-offset-x"))
            MegaChunk.MEGA_CHUNK_OFFSET_X = ThreadLocalRandom.current().nextInt()
        if (config.getBoolean("megachunk-random-offset-z"))
            MegaChunk.MEGA_CHUNK_OFFSET_Z = ThreadLocalRandom.current().nextInt()
    }

    override fun onDisable() {
        HandlerList.unregisterAll(this)
        redis.close()
    }

    override fun prepareAcf(commandManager: PaperCommandManager): PaperCommandManager {
        fun messageType(type: MessageType) {
            for (colour in BukkitChatColor.values()) {
                commandManager.setFormat(type, colour.ordinal, colour)
            }
        }
        messageType(MessageType.ERROR)
        messageType(MessageType.INFO)
        messageType(MessageType.SYNTAX)
        messageType(MessageType.HELP)

        val acfLangFile = File(commandManager.plugin.dataFolder, "acf-lang-en.yml")
        copyResource("acf-lang-en.yml", acfLangFile)

        commandManager.locales.loadYamlLanguageFile(acfLangFile, Locale.ENGLISH)

        return commandManager
    }

    override fun prepareDatabase(
        databaseName: String,
        username: String,
        password: String,
        hostAndPort: String,
        plugin: Plugin
    ): Database {
        return plugin.createHikariDatabase(
            username.ifEmpty { (config.getString("sql-username") ?: throw IllegalArgumentException("user cannot be null")) },
            password.ifEmpty { (config.getString("sql-password") ?: throw IllegalArgumentException("password cannot be null")) },
            (config.getString("sql-db-prefix") ?: "") + databaseName,
            hostAndPort.ifEmpty { (config.getString("sql-host-and-port")
                ?: throw IllegalArgumentException("host and port cannot be null")) }
        )
    }

    companion object {
        /**
         * Gets the current instance of this plugin.
         *
         * @return current instance of this plugin.
         */
        val instance: MinevictusUtilsSpigot
            get() = getPlugin(MinevictusUtilsSpigot::class.java)
    }
}